projects/com.oracle.truffle.llvm.types/src/com/oracle/truffle/llvm/types/vector/LLVMDoubleVector.java

```diff
@@ -49,48 +55,90 @@ public static LLVMDoubleVector fromDoubleArray(LLVMAddress target, double[] vals
         return new LLVMDoubleVector(target, vals.length);
     }
 
-    @Override
-    public int getElementByteSize() {
-        return DOUBLE_SIZE;
-    }
-
-    @Override
-    protected LLVMVector<Double> create(LLVMAddress addr, int length) {
+    private static LLVMDoubleVector create(LLVMAddress addr, int length) {
         return new LLVMDoubleVector(addr, length);
     }
 
-    @Override
-    public Double getValue(LLVMAddress addr) {
-        return LLVMMemory.getDouble(addr);
-    }
-
-    @Override
-    public void setValue(LLVMAddress addr, Double value) {
-        LLVMMemory.putDouble(addr, value);
-    }
-
     public static LLVMDoubleVector createDoubleVector(LLVMAddress addr, int size) {
         return new LLVMDoubleVector(addr, size);
     }
 
+    public double[] getValues() {
+        double[] values = new double[nrElements];
+        for (int i = 0; i < values.length; i++) {
+            values[i] = getValue(i);
+        }
+        return values;
+    }
+
     public LLVMDoubleVector add(LLVMAddress addr, LLVMDoubleVector right) {
-        return performOperation(addr, right, (a, b) -> a + b);
+        LLVMAddress currentAddr = addr;
+        for (int i = 0; i < nrElements; i++) {
```

Comment:
Removing the lambdas makes sense. However, we duplicated lots of code now. Could we instead use anonymous classes or use some other abstraction?